### DISCUSSION TITLE
[Architecture discussion]How do we gain state overview in a relatively large codebase

### DISCUSSION THREAD

**Discussion by zoeyzou** [pmndrs/jotai#2167] [Q&A]
So with the other state management, they do distinguish global state vs local state; so for global state, usually in the root level you will see patterns like this:

```
<ThemeContext>
  <QueryClient>
    <AuthenticationContext>
      <StepContext>
          ...
      </StepContext>
    </AuthenticationContext>
  </QueryClient>
</ThemeContext>
```

This way as a developer coming into the codebase, you can easily know what are available and how the app is composed. But with the atomic approach of Jotai, I wonder how do we ensure the DX and provide an overview on what's there since everything is globally accessible, and how do we make sure that the decision of spawning a state is somehow guided/gated?

Would love to hear people's opinion and idea.

**Comment by dai-shi** [comment]
Think atoms like functions. How do you define function in a large code base?
I would use directory structure and modules (import/export) and sometimes closures (atom creators).

**Reply by zoeyzou** [reply]
Hi @dai-shi , thanks for answering. So in the codebase I'm working on, we are using "Scream" structure, aka we put things that should be together as a group in our feature folder. But there might be some level of circular dependency. So right now we are mostly using context, we have a few layers of global states that live in their own context, and it's very obvious what is available globally. Coming into the codebase, developers know where to get some of the states.

However, if each atoms are in their own feature folder (directory structure), and there's no central place that knows about it (they are not mentioned in the root any anywhere else apart from where it's used), how do we ensure the DX and overviews? Although this question might be a bit anti-pattern for Jotai, I just wonder if people already have some smarter solution
